daemon/src/lib.rs

```diff
@@ -0,0 +1,655 @@
+/*
+Copyright (c) 2022 VMware, Inc.
+SPDX-License-Identifier: MIT
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+ */
+
+// Derived from lib/daemon-unix.c in Open vSwitch, with the following license:
+/*
+ * Copyright (c) 2008, 2009, 2010, 2011, 2012, 2013, 2015 Nicira, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Utilities for a daemon to detach and monitor itself.
+//!
+//! In Unix-like environments, it's traditional for daemon processes to be able to **detach**
+//! itself from the current session and run, isolated, in the background.  It's also useful to have
+//! a mechanism for a daemon to **monitor** itself and automatically restart if it dies due to a
+//! signal that indicates an error (such as SIGABRT or SIGSEGV).  These days, both of these
+//! functions are ones that can be implemented externally (e.g. by `systemd`), but it can still be
```

Comment:
`init` only sort-of did that. It would restart anything listed in `/etc/inittab`, but that was a very limited set of processes. Usually it was limited to `getty` run on a few terminals and maybe an `xdm` for a graphical session. For general-purpose daemons, instead it just ran a script (that ran a lot more scripts) when runlevels changed. Those scripts then had to arrange to start and monitor themselves.